File: ui/shared/apollo/InstAccess.js
/*
 * Copyright (C) 2021 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import getCookie from '@instructure/get-cookie'

class InstAccess {
  // This class is intended to be used in combination with our
  // ApolloClient when talking to the API Gateway rather than
  // Canvas directly.  It should provide an in-memory place to store
  // the access token (JWT) we need for gateway access without needing to
  // put it in local storage or an easy to poke at global variable.
  //  When we substitute the `gatewayAuthenticatedFetch` as the "fetch"
  // option for an Apollo HttpLink, it will add an inst access token
  // to the header (fetching one if necessary), and should refresh
  // the token and retry once if it ever receives a 4xx (which we
  // would expect in the case when an InstAccess token has passed it's
  // expiration window).
  constructor(options = {}) {
    this.instAccessToken = options.token || null
    this.fetchImpl =
      options.preferredFetch ||
      (async (uri, opts) => {
        return fetch(uri, opts)
      })
  }

}

export default InstAccess

File: gems/inst_access/lib/inst_access.rb
# frozen_string_literal: true

#
# Copyright (C) 2021 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

require 'active_support'
require 'json/jwt'
require 'inst_access/errors'
require 'inst_access/config'
require 'inst_access/token'

module InstAccess
  class << self
    # signing_key is required.  if you are going to be producing (and therefore
    # signing) tokens, this needs to be an RSA private key.  if you're just
    # consuming tokens, it can be the RSA public key corresponding to the
    # private key that signed them.
    # encryption_key is only required if you are going to be producing tokens.
    def configure(signing_key:, encryption_key: nil)
      @config = Config.new(signing_key, encryption_key)
    end

    # set a configuration only for the duration of the given block, then revert
    # it.  useful for testing.
    def with_config(signing_key:, encryption_key: nil)
      old_config = @config
      configure(signing_key: signing_key, encryption_key: encryption_key)
      yield
    ensure
      @config = old_config
    end

    def config
      @config || raise(ConfigError, "InstAccess is not configured!")
    end
  end
end
